requirements.py

```diff
@@ -16,29 +16,162 @@
+
+                elif line.startswith("-"):
+                    # another special command we don't recognize
+                    stuff.append(line)
+
+                else:
+                    # ordenary requirement, similary to them used in setup.py
+                    rnormal.append(line)
+    except IOError:
+        print("Warning: could not parse requirements file '{}'!", file=sys.stderr)
+
+    return rdev, rnormal, stuff
+
 if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description="Calculates requirements for different purposes",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter
+    )
+    parser.add_argument(
+        "-l", "--level",
+        choices=["min", "pypi", "dev"],
+        default="pypi",
+        help="Specifies desired requirements level. 'min' requests the minimal requirement that is specified, 'pypi' requests the maximum version that satisfies the constrains and is available in PyPi. 'dev' includes experimental developer versions for VCSs."
+    )
+    parser.add_argument(
+        "-e", "--extras",
+        default="",
+        help="Comma separated list of extras.",
+        type=parse_set
+    )
+    args = parser.parse_args()
+
+    result = dict()
+    requires = []
+    stuff = []
+    if args.level == "dev":
+        result, requires, stuff = parse_pip_file("requirements-devel.txt")
+
     with mock.patch.object(setuptools, 'setup') as mock_setup:
-        import setup  # pylint: disable=F401
+        import setup
+        assert setup  # silence warning about unused imports
 
     # called arguments are in `mock_setup.call_args`
-    args, kwargs = mock_setup.call_args
-    install_requires = kwargs.get('install_requires', [])
-
-    for pkg in pkg_resources.parse_requirements(install_requires):
-        if len(pkg.specs):
-            if pkg.specs[0][0] == '>=':
-                print("{0.project_name}=={0.specs[0][1]}".format(pkg))
-            elif pkg.specs[0][0] == '>':
-                print(
-                    "{0.project_name} specify exact minimal version using "
-                    "'>=' instead of '>'.".format(pkg), file=sys.stderr)
+    mock_args, mock_kwargs = mock_setup.call_args
+    requires = mock_kwargs.get('install_requires', [])
+
+    requires_extras = mock_kwargs.get('extras_require', {})
+    for e in args.extras:
+        if e in requires_extras:
+            requires.extend(requires_extras[e])
+
+    for pkg in pkg_resources.parse_requirements(requires):
+        # skip things we already know
+        # FIXME be smarter about merging things
+        if pkg.key in result:
+            continue
+
+        # check if package should be included
+        include = False
+        if not pkg.extras:
+            include = True
+        else:
+            for e in pkg.extras:
+                if e in args.extras:
+                    include = True
+
+        if include:
+            # remove extra marker
+            pkg.extras = set()
```

Comment:
This might be important for certain packages. (e.g. `SQLAlchemy-Utils[encrypted]`).